parachain/pallets/collab-ai/aiusd-convertor/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+	traits::{
+		tokens::{
+			fungibles::{Inspect as FsInspect, Mutate as FsMutate},
+			Fortitude, Precision, Preservation,
+		},
+		StorageVersion,
+	},
+	transactional,
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_runtime::traits::{CheckedDiv, CheckedMul};
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	pub type InspectFungibles<T> = pallet_assets::Pallet<T>;
+	/// Balance type alias for balances of assets that implement the `fungibles` trait.
+	pub type AssetBalanceOf<T> =
+		<InspectFungibles<T> as FsInspect<<T as frame_system::Config>::AccountId>>::Balance;
+	/// Type alias for Asset IDs.
+	pub type AssetIdOf<T> =
+		<InspectFungibles<T> as FsInspect<<T as frame_system::Config>::AccountId>>::AssetId;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_assets::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		// This is not a treasury account
+		// Balance of all types in this account record the
+		// available stable coin for AIUSD to switch back
+		type ConvertingFeeAccount: Get<Self::AccountId>;
+
+		// Declare the asset id of AIUSD
+		type AIUSDAssetId: Get<AssetIdOf<Self>>;
+
+		/// The privileged origin to enable/disable AIUSD switch
+		type ManagerOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+	}
+
+	// Asset Id => ratio of system exchange rate for AIUSD to other token in 10^18
+	// e.g.
+	// (1 USDT) = 10^6 in system
+	// (1 AIUSD) = 10^18 in system
+	// Value of StorageMap n = 10^(-12) * 10^(18) = 10^(6),
+	// which means (1 AIUSD) * n = (10^18) * (1 USDT) in system balance when converting.
+	#[pallet::storage]
+	#[pallet::getter(fn enabled_tokens)]
+	pub type EnabledTokens<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, AssetBalanceOf<T>, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidAssetId,
+		AssetNotEnabled,
+		Overflow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// AIUSD minted with other eligible stable token locked
+		AIUSDCreated {
+			beneficiary: T::AccountId,
+			aiusd_amount: AssetBalanceOf<T>,
+			asset_id: AssetIdOf<T>,
+			asset_amount: AssetBalanceOf<T>,
+		},
+		/// AIUSD burned with other eligible stable token released
+		AIUSDDestroyed {
+			beneficiary: T::AccountId,
+			aiusd_amount: AssetBalanceOf<T>,
+			asset_id: AssetIdOf<T>,
+			asset_amount: AssetBalanceOf<T>,
+		},
+		AssetEnabled {
+			asset_id: AssetIdOf<T>,
+			decimal_ratio: AssetBalanceOf<T>,
+		},
+		AssetDisabled {
+			asset_id: AssetIdOf<T>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		// Lock target_asset_id token and mint AIUSD
```

Comment:
Switch to ConvertingFeeAccount to ConvertingPool.

We use lock since we want to make sure that the bridge on remote chains (ethereum, bsc etc.) have enough token to support the swap. Otherwise we will lose track of distribution of assets.

Reply:
Do you mean we need something to track the liquidity so that we can't e.g. bridge out more assets than what exists in the pool?

Then I still think pallet-asset-conversion is a better solution, when bridging in, you are actually adding liquidity to the pool and swapping the for AIUSD, it's also guaranteed that when all assets are drained, you can't do the swap anymore.

Actually it should be used for our chain bridge too 🤔